### ISSUE TITLE
When using func-names = never, allow names for recursive use

### ISSUE THREAD

**Issue by kewisch** [eslint/eslint#6616]
**What version of ESLint are you using?** 2.13.0

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:** `"func-names": [2, "never"]`

**The problem you want to solve**
I would like to never use function names, because in Mozilla code the names in stacks are guessed pretty well nowadays. There are some cases where I still need them though, namely when referencing the function name within for a recursive call.

Example:
```javascript
setTimeout(function postpone() {
    let thing = getThing();
    if (!thing) {
        setTimeout(postpone, 0);
    } else {
        complete();
    }
}, 0);
```

**Your take on the correct solution to problem.**
I would suggest to add an keyword option `recursive` or similar, setting it to `true` would allow use of function names if referenced from within. The default should be `true`.

**Comment by alberto** [comment]
Seems reasonable to me.

**Comment by btmills** [comment]
Right now `func-names` is either "always" or "never". What about making it "always/recursive/never"? That would be a non-breaking enhancement and avoid a boolean object option.

**Comment by mysticatea** [comment]
I think that we can address this issue as a false positive.

- [prefer-arrow-callback](http://eslint.org/docs/rules/prefer-arrow-callback) skips functions if the name of the function is used. This is similar to that.
- This fix would reduce warnings. (it's `patch` thing.)

**Comment by alberto** [comment]
Good point. @btmills do you see any value in having it as a separate option?

**Comment by btmills** [comment]
@mysticatea's argument makes sense. Right now "never" has the side effect of forbidding recursion, which is (supposed to be) unrelated to this rule. I'd support classifying this as a bug and excepting recursive named function expressions from "never".

**Comment by alberto** [comment]
I'm working on this.

**Comment by mysticatea** [comment]
Thank you!
